File: tests/compat/clients/cycle.js
/*
 * Le cycle de vie d'un client officiel, exerce par `@elastic/elasticsearch` 8.x.
 *
 * Lance dans un conteneur par `tests/compat/tests_clients.py`, contre l'URL
 * passee en argument. N'importe que le client officiel installe depuis npm :
 * le code ci-dessous est ecrit ici, la bibliotheque qu'il exerce ne l'est pas.
 *
 * Meme format de sortie que les deux autres batteries :
 * `CAS <nom> <PASS|FAIL> <detail>`.
 */
'use strict'

const { Client, errors, events } = require('@elastic/elasticsearch')

const URL = process.argv[2] || 'http://localhost:9200'
const INDEX = 'cycle-javascript'
const CAS = []

function cas (nom, f) { CAS.push([nom, f]) }
function exige (condition, message) { if (!condition) throw new Error(message) }

// ---------------------------------------------------------------------------
// 1. Ce que le client fait avant qu'on lui demande quoi que ce soit
// ---------------------------------------------------------------------------

cas('decouverte_version', async (es) => {
  const info = await es.info()
  const numero = info.version.number
  exige(numero.split('.')[0] === '8', `version majeure ${numero}, le client 8.x exige 8`)
  exige(info.tagline === 'You Know, for Search', info.tagline)
  exige(Boolean(info.cluster_name), 'cluster_name vide')
  return numero
})

cas('entete_produit', async (es) => {
  // Sans `x-elastic-product`, le client 8.x leve `ProductNotSupportedError`
  // et aucun appel ne passe. Verifie sur trois formes de reponse.
  const vus = {}
  vus['/'] = (await es.info({}, { meta: true })).headers['x-elastic-product']
  vus._search = (await es.search({ index: '*', size: 0 }, { meta: true })).headers['x-elastic-product']
  vus._bulk = (await es.bulk({
    operations: [{ index: { _index: INDEX, _id: 'entete' } }, { n: 1 }],
    refresh: true
  }, { meta: true })).headers['x-elastic-product']
  const manquants = Object.keys(vus).filter((k) => vus[k] !== 'Elasticsearch')
  exige(manquants.length === 0, `en-tete absent ou faux sur ${manquants} (${JSON.stringify(vus)})`)
  return 'Elasticsearch sur ' + Object.keys(vus).join(', ')
})

cas('negociation_compression', async (_es, url) => {
  // `compression: true` gzippe le corps de la requete. Un serveur qui ne sait
  // pas le lire rend un 400 sur un JSON qu'il juge invalide.
  const gz = new Client({ node: url, compression: true })
  try {
    const lignes = []
    for (let i = 0; i < 200; i++) {
      lignes.push({ index: { _index: INDEX, _id: `gz${i}` } })
      lignes.push({ texte: `document ${i} `.repeat(40) })
    }
    const rep = await gz.bulk({ operations: lignes, refresh: true })
    exige(rep.errors === false, JSON.stringify(rep.items && rep.items[0]))
    const trouves = await gz.search({ index: INDEX, query: { match: { texte: 'document' } }, size: 0 })
    exige(trouves.hits.total.value >= 200, JSON.stringify(trouves.hits.total))
    return '200 documents envoyes en gzip, relus'
  } finally {
    await gz.close()
  }
})

cas('sniffing', async (_es, url) => {
  // `sniffOnStart` demande `GET /_nodes/_all/http` et remplace le pool par ce
  // qu'il rend. Le client JavaScript ne leve pas si le sniff echoue : il emet
  // un evenement. Le lire est donc la seule facon de savoir laquelle des deux
  // reponses on a eue — sans lui, un refus se lirait « sniffing tenu ».
  const sniff = new Client({ node: url, sniffOnStart: true, sniffEndpoint: '_nodes/_all/http' })
  let echec = null
  let vu = false
  sniff.diagnostic.on(events.SNIFF, (err) => { vu = true; echec = err })
  try {
    await new Promise((resolve) => setTimeout(resolve, 500))
    const info = await sniff.info()
    exige(Boolean(info.version.number), JSON.stringify(info))
    exige(vu, "le client n'a pas sniffe du tout")
    if (echec == null) return 'sniffing tenu, le client a garde la main'
    exige(String(echec.message || '').trim().length > 0, `${echec.name} sans message : un refus muet`)
    return `refuse proprement, client toujours utilisable (${echec.name}: ${String(echec.message).slice(0, 100)})`
  } finally {
    await sniff.close()
  }
})

// ---------------------------------------------------------------------------
// 2. Ce que le client fait des erreurs
// ---------------------------------------------------------------------------

cas('erreurs_typees', async (es) => {
  const vus = []

  try {
    await es.get({ index: INDEX, id: 'jamais-ecrit' })
    throw new Error('un document absent aurait du lever ResponseError')
  } catch (e) {
    exige(e instanceof errors.ResponseError, `${e.name} au lieu de ResponseError`)
    exige(e.statusCode === 404, `statut ${e.statusCode}`)
    vus.push('ResponseError(404)')
  }

  try {
    await es.search({ index: INDEX, query: { pas_une_clause: {} } })
    throw new Error('une clause inconnue aurait du lever ResponseError')
  } catch (e) {
    exige(e instanceof errors.ResponseError, `${e.name} au lieu de ResponseError`)
    exige(e.statusCode === 400, `statut ${e.statusCode}`)
    exige(Boolean(e.body.error.type), JSON.stringify(e.body))
    exige(Boolean(e.body.error.reason), JSON.stringify(e.body))
    vus.push(`ResponseError(400, ${e.body.error.type})`)
  }

  await es.index({ index: INDEX, id: 'conflit', document: { n: 1 }, refresh: true })
  try {
    await es.index({ index: INDEX, id: 'conflit', document: { n: 2 }, if_seq_no: 99999, if_primary_term: 1 })
    throw new Error('un `if_seq_no` perime aurait du lever')
  } catch (e) {
    exige(e.statusCode === 409, `statut ${e.statusCode}`)
    exige(e.body.error.type === 'version_conflict_engine_exception', JSON.stringify(e.body))
    vus.push('ResponseError(409, version_conflict_engine_exception)')
  }

  // L'option `ignore` transforme un statut attendu en reponse normale : c'est
  // ce que fait tout code qui teste l'existence d'un document.
  const absent = await es.get({ index: INDEX, id: 'toujours-pas' }, { ignore: [404] })
  exige(absent.found === false, JSON.stringify(absent))
  vus.push('ignore:[404] rend le corps')

  return vus.join(', ')
})

// ---------------------------------------------------------------------------
// 3. Les helpers
// ---------------------------------------------------------------------------

cas('helpers_bulk', async (es) => {
  // Le helper decoupe, envoie, retente, et compte. `onDrop` est appele pour
  // chaque document refuse : un helper qui avalerait un rejet rendrait un
  // index incomplet en silence.
  const documents = []
  for (let i = 0; i < 1500; i++) documents.push({ rang: i, flux: true })
  const jetes = []
  const resultat = await es.helpers.bulk({
    datasource: documents,
    flushBytes: 200000,
    refreshOnCompletion: INDEX,
    onDrop (doc) { jetes.push(doc) },
    onDocument (doc) { return { index: { _index: INDEX, _id: `b${doc.rang}` } } }
  })
  exige(jetes.length === 0, `${jetes.length} documents jetes`)
  exige(resultat.successful === 1500, `${resultat.successful} indexes`)
  exige(resultat.failed === 0, `${resultat.failed} echecs`)
  const compte = await es.count({ index: INDEX, query: { term: { flux: true } } })
  exige(compte.count === 1500, `${compte.count} documents relus`)
  return `1500 documents indexes en ${resultat.total} operations, 0 jete`
})

cas('helpers_scroll', async (es) => {
  // `scrollSearch` deroule un `scroll` page par page et ferme le contexte.
  // Chaque document doit sortir une fois et une seule.
  const vus = new Set()
  let pages = 0
  for await (const reponse of es.helpers.scrollSearch({
    index: INDEX,
    query: { term: { flux: true } },
    size: 137
  })) {
    pages += 1
    for (const hit of reponse.documents) vus.add(hit.rang)
  }
  exige(vus.size === 1500, `${vus.size} documents distincts sur 1500, en ${pages} pages`)
  return `1500 documents deroules en ${pages} pages, sans doublon`
})

// ---------------------------------------------------------------------------

async function main () {
  const es = new Client({ node: URL })
  if (await es.indices.exists({ index: INDEX })) {
    await es.indices.delete({ index: INDEX })
  }
  await es.indices.create({
    index: INDEX,
    mappings: {
      properties: {
        rang: { type: 'integer' },
        flux: { type: 'boolean' },
        texte: { type: 'text' },
        n: { type: 'integer' }
      }
    }
  })
  let rates = 0
  for (const [nom, f] of CAS) {
    try {
      const detail = await f(es, URL)
      console.log(`CAS ${nom} PASS ${detail}`)
    } catch (e) {
      console.log(`CAS ${nom} FAIL ${e.name}: ${String(e.message).split('\n')[0]}`)
      for (const ligne of String(e.stack || e).split('\n')) console.log(`    | ${ligne}`)
      rates += 1
    }
  }
  await es.indices.delete({ index: INDEX }, { ignore: [404] })
  await es.close()
  console.log(`CYCLE javascript ${CAS.length - rates}/${CAS.length}`)
  process.exit(rates ? 1 : 0)
}

main().catch((e) => {
  console.log(`CAS demarrage FAIL ${e.name}: ${String(e.message).split('\n')[0]}`)
  console.log(`CYCLE javascript 0/${CAS.length}`)
  process.exit(1)
})

File: tests/compat/clients/cycle.py
#  Le cycle de vie d'un client officiel, exerce par `elasticsearch-py` 8.x.
#
#  Lance dans un conteneur par `tests/compat/tests_clients.py`, contre l'URL
#  passee en argument. N'importe **que** le client officiel installe depuis
#  PyPI : le code ci-dessous est ecrit ici, la bibliotheque qu'il exerce ne
#  l'est pas.
#
#  Chaque cas imprime une ligne `CAS <nom> <PASS|FAIL> <detail>` — un format
#  volontairement pauvre, pour que les trois langages rendent la meme chose.
import sys
import traceback

from elasticsearch import (
    ApiError,
    BadRequestError,
    ConflictError,
    Elasticsearch,
    NotFoundError,
    helpers,
)

URL = sys.argv[1] if len(sys.argv) > 1 else "http://localhost:9200"
INDEX = "cycle-python"

CAS = []


def cas(nom):
    def decore(f):
        CAS.append((nom, f))
        return f

    return decore


# ---------------------------------------------------------------------------
# 1. Ce que le client fait avant qu'on lui demande quoi que ce soit
# ---------------------------------------------------------------------------


@cas("decouverte_version")
def decouverte_version(es):
    """Le premier appel de tout client : la version, le nom du cluster, la
    signature. Un numero que le client juge incompatible, et il refuse de
    parler au serveur — avant meme la premiere recherche."""
    info = es.info()
    numero = info["version"]["number"]
    majeur = int(numero.split(".")[0])
    assert majeur == 8, f"version majeure {majeur}, le client 8.x exige 8"
    assert info["tagline"] == "You Know, for Search", info["tagline"]
    assert info["cluster_name"], "cluster_name vide"
    assert info["version"]["build_flavor"], "build_flavor absent"
    return numero


@cas("entete_produit")
def entete_produit(es):
    """`X-elastic-product: Elasticsearch`. Sans lui, le client 8.x leve
    `UnsupportedProductError` et **aucun** appel ne passe : c'est la premiere
    chose qui separe « un serveur qui repond du JSON » d'« un Elasticsearch ».
    Il est verifie sur trois formes de reponse, pas seulement sur la racine."""
    vus = {}
    vus["/"] = es.info().meta.headers.get("x-elastic-product")
    vus["_search"] = es.search(index="*", size=0).meta.headers.get("x-elastic-product")
    vus["_bulk"] = es.bulk(
        operations=[{"index": {"_index": INDEX, "_id": "entete"}}, {"n": 1}],
        refresh=True,
    ).meta.headers.get("x-elastic-product")
    manquants = [route for route, v in vus.items() if v != "Elasticsearch"]
    assert not manquants, f"en-tete absent ou faux sur {manquants} ({vus})"
    return "Elasticsearch sur " + ", ".join(vus)


@cas("negociation_compression")
def negociation_compression(es_url):
    """`http_compress=True` : le client gzippe **le corps de sa requete** et
    pose `Content-Encoding: gzip`. Un serveur qui ne sait pas le lire rend un
    400 sur un JSON qu'il juge invalide — donc un echec qui ne nomme pas sa
    cause. Le lot est volontairement gros : c'est la ou un client active la
    compression."""
    with Elasticsearch(es_url, http_compress=True) as gz:
        docs = [{"_index": INDEX, "_id": f"gz{i}", "texte": f"document {i} " * 40} for i in range(200)]
        ok, echecs = helpers.bulk(gz, docs, refresh=True)
        assert ok == 200 and not echecs, f"{ok} indexes, {echecs}"
        trouves = gz.search(index=INDEX, query={"match": {"texte": "document"}}, size=0)
        assert trouves["hits"]["total"]["value"] >= 200, trouves["hits"]["total"]
    return "200 documents envoyes en gzip, relus"


@cas("sniffing")
def sniffing(es_url):
    """Le sniffing demande la liste des noeuds (`GET /_nodes/_all/http`) et
    remplace le pool de connexions par ce qu'elle rend. ferrite est mono-noeud :
    le tenir ou le refuser sont deux reponses acceptables — **se taire** n'en
    est pas une. Le cas exige donc l'un des deux, nomme."""
    try:
        with Elasticsearch(es_url, sniff_on_start=True, min_delay_between_sniffing=0) as sniff:
            info = sniff.info()
            assert info["version"]["number"], info
        return "sniffing tenu, le client a garde la main"
    except Exception as e:  # noqa: BLE001 — c'est le refus qu'on mesure
        message = str(e)
        assert message.strip(), f"{type(e).__name__} sans message : un refus muet"
        assert "no handler" in message or "ferrite ne supporte pas" in message or "sniff" in message.lower(), (
            f"refus illisible : {type(e).__name__}: {message[:200]}"
        )
        return f"refuse proprement ({type(e).__name__}: {message[:120]})"


# ---------------------------------------------------------------------------
# 2. Ce que le client fait des erreurs
# ---------------------------------------------------------------------------


@cas("erreurs_typees")
def erreurs_typees(es):
    """Un client officiel ne rend pas un code de statut : il leve une classe.
    Le mapping statut -> classe se fait sur le **corps** de l'erreur, donc il
    exige le format d'ES — `error.type`, `error.reason`, `status`."""
    vus = []

    try:
        es.get(index=INDEX, id="jamais-ecrit")
        raise AssertionError("un document absent aurait du lever NotFoundError")
    except NotFoundError as e:
        assert e.status_code == 404, e.status_code
        # Un document absent n'est pas une erreur d'API : ES rend le meme corps
        # qu'un document trouve, `found: false`, sous un 404.
        assert e.body.get("found") is False, e.body
        assert e.body["_id"] == "jamais-ecrit", e.body
        vus.append("NotFoundError(404, found:false)")

    try:
        es.search(index=INDEX, query={"pas_une_clause": {}})
        raise AssertionError("une clause inconnue aurait du lever BadRequestError")
    except BadRequestError as e:
        assert e.status_code == 400, e.status_code
        assert e.body["error"]["type"], e.body
        assert e.body["error"]["reason"], e.body
        vus.append(f"BadRequestError(400, {e.body['error']['type']})")

    es.index(index=INDEX, id="conflit", document={"n": 1}, refresh=True)
    try:
        es.index(index=INDEX, id="conflit", document={"n": 2}, if_seq_no=99999, if_primary_term=1)
        raise AssertionError("un `if_seq_no` perime aurait du lever ConflictError")
    except ConflictError as e:
        assert e.status_code == 409, e.status_code
        assert e.body["error"]["type"] == "version_conflict_engine_exception", e.body
        vus.append("ConflictError(409)")

    try:
        es.indices.create(index=INDEX)
        raise AssertionError("un index deja la aurait du lever une ApiError")
    except ApiError as e:
        assert e.status_code == 400, e.status_code
        assert e.body["error"]["type"] == "resource_already_exists_exception", e.body
        vus.append("ApiError(400, resource_already_exists_exception)")

    return ", ".join(vus)


# ---------------------------------------------------------------------------
# 3. Les helpers — le code que personne ne reecrit
# ---------------------------------------------------------------------------


@cas("helpers_bulk")
def helpers_bulk(es):
    """`helpers.bulk` decoupe, envoie, et **leve** sur le premier document
    refuse. Les deux moities comptent : un helper qui avale un rejet rendrait
    un index incomplet en silence."""
    docs = [{"_index": INDEX, "_id": f"b{i}", "rang": i} for i in range(1500)]
    ok, echecs = helpers.bulk(es, docs, chunk_size=250, refresh=True)
    assert ok == 1500, ok
    assert not echecs, echecs
    total = es.count(index=INDEX, query={"range": {"rang": {"gte": 0}}})["count"]
    assert total == 1500, total

    mauvais = [{"_index": INDEX, "_id": "b0", "_op_type": "create", "rang": 0}]
    try:
        helpers.bulk(es, mauvais)
        raise AssertionError("un `create` sur un id existant aurait du lever")
    except helpers.BulkIndexError as e:
        assert e.errors, e
    return "1500 documents en 6 lots, et le rejet leve"


@cas("helpers_streaming_bulk")
def helpers_streaming_bulk(es):
    """`streaming_bulk` rend un resultat **par document**, dans l'ordre — c'est
    ce qui permet a un appelant de rattacher un echec a sa ligne."""
    def source():
        for i in range(600):
            yield {"_index": INDEX, "_id": f"s{i}", "rang": i, "flux": True}

    rendus = []
    for succes, detail in helpers.streaming_bulk(es, source(), chunk_size=100, refresh=True):
        assert succes, detail
        rendus.append(next(iter(detail.values()))["_id"])
    assert len(rendus) == 600, len(rendus)
    assert rendus == [f"s{i}" for i in range(600)], "l'ordre des resultats n'est pas celui des documents"
    return "600 resultats, un par document, dans l'ordre"


@cas("helpers_parallel_bulk")
def helpers_parallel_bulk(es):
    """Le meme, sur quatre fils : ce que fait un import un peu presse."""
    docs = ({"_index": INDEX, "_id": f"p{i}", "rang": i} for i in range(800))
    combien = 0
    for succes, detail in helpers.parallel_bulk(es, docs, thread_count=4, chunk_size=100):
        assert succes, detail
        combien += 1
    assert combien == 800, combien
    return "800 documents par 4 fils"


@cas("helpers_scan")
def helpers_scan(es):
    """`helpers.scan` est le code de tout export : il ouvre un `scroll`, deroule
    toutes les pages et ferme le contexte. Chaque document doit sortir **une
    fois et une seule**."""
    es.indices.refresh(index=INDEX)
    vus = [d["_id"] for d in helpers.scan(es, index=INDEX, query={"query": {"term": {"flux": True}}}, size=97)]
    assert len(vus) == 600, len(vus)
    assert len(set(vus)) == 600, "des documents sortent deux fois"

    tries = [
        d["_source"]["rang"]
        for d in helpers.scan(
            es,
            index=INDEX,
            query={"query": {"match_all": {}}, "sort": ["rang"]},
            preserve_order=True,
            size=50,
        )
        if "rang" in d["_source"]
    ]
    assert tries == sorted(tries), "preserve_order ne preserve pas l'ordre"
    return f"600 documents deroules sans doublon, {len(tries)} en ordre conserve"


# ---------------------------------------------------------------------------


def main():
    with Elasticsearch(URL) as es:
        if es.indices.exists(index=INDEX):
            es.indices.delete(index=INDEX)
        es.indices.create(
            index=INDEX,
            mappings={
                "properties": {
                    "rang": {"type": "integer"},
                    "flux": {"type": "boolean"},
                    "texte": {"type": "text"},
                    "n": {"type": "integer"},
                }
            },
        )
        rate = 0
        for nom, f in CAS:
            try:
                # Les cas qui construisent leur propre client prennent l'URL,
                # les autres celui qui est deja ouvert.
                detail = f(URL if f.__code__.co_varnames[0] == "es_url" else es)
                print(f"CAS {nom} PASS {detail}")
            except Exception:  # noqa: BLE001 — un cas rouge n'arrete pas la batterie
                trace = traceback.format_exc().strip().splitlines()
                print(f"CAS {nom} FAIL {trace[-1]}")
                for ligne in trace:
                    print(f"    | {ligne}")
                rate += 1
        es.options(ignore_status=404).indices.delete(index=INDEX)
    print(f"CYCLE python {len(CAS) - rate}/{len(CAS)}")
    return 1 if rate else 0


if __name__ == "__main__":
    sys.exit(main())
